pkg/personname/info.go

```diff
@@ -0,0 +1,202 @@
+func (info Info) WithoutNullSeparators() Info {
+	info.NoNullSeparators = true
+	info.Alphabetic.NoNullSeparators = true
+	info.Ideographic.NoNullSeparators = true
+	info.Phonetic.NoNullSeparators = true
+
+	return info
+}
+
+// dcmRemoveNullStrings removes null group strings from the list of groups to be
+// rendered.
+func (info Info) dcmRemoveNullStrings(groupStrings []string) []string {
+	// If we are not removing null separators, return immediately.
+	if !info.NoNullSeparators {
+		return groupStrings
+	}
+
+	// If we are removing trailing emtpy, we are going to check if the last group is
+	// empty, and if it is, trim it off the end until we hit a non-empty slice.
+	//
+	// If we were to move front-to-back instead of back-to-front, we would get a bad
+	// result if the middle group is blank, but the third group is not, as we would
+	// bail on the middle group, and not end up rendering the last.
+	//
+	// 'Harry^Potter==Phonetic^Spelling is a valid PN value, which we need to account
+	// for.
+
+	// Start with the sliceOut point containing all three groups
+	sliceOut := 3
+	// Iterate backwards over the strings
+	for i := 2; i >= 0; i-- {
+		// If the string is blank, shift the slice out back 1 place.
+		if groupStrings[i] == "" {
+			sliceOut = i
+		} else {
+			// Otherwise, if it is not blank, we have to render all remaining
+			// separators, even if their values are null, so bail.
+			break
+		}
+	}
+
+	// Trim the groups by getting a slice of our slice using the sliceOut value.
+	groupStrings = groupStrings[0:sliceOut]
+	return groupStrings
+}
+
+// DCM returns the serialized DICOM representation of the PN value, in
+// '[Alphabetic]=[Ideographic]=[Phonetic]' format.
+func (info Info) DCM() string {
+	// Convert the groups into their formatted string representations.=
+	groupStrings := []string{
+		info.Alphabetic.DCM(), info.Ideographic.DCM(), info.Phonetic.DCM(),
+	}
+
+	// Remove groups based on the trailing null setting and the formatted group strings.
+	groupStrings = info.dcmRemoveNullStrings(groupStrings)
+
+	// Join the remaining groups with '='
+	return strings.Join(groupStrings, groupSep)
+}
+
+// IsEmpty returns whether the PN value contains any actual information. This method
+// ignores separator, so both '' and '^^^^=^^^^=^^^^' would return true.
+func (info Info) IsEmpty() bool {
+	return info.Alphabetic.IsEmpty() &&
+		info.Ideographic.IsEmpty() &&
+		info.Phonetic.IsEmpty()
+}
+
+// Parse PN dicom value into a personname.Info value.
+func Parse(valueString string) (Info, error) {
+	groups := strings.Split(valueString, groupSep)
+
+	// If there are more than three groups, then the value does not conform to the DICOM
+	// spec.
+	if len(groups) > 3 {
+		return Info{}, newErrTooManyGroups(len(groups))
+	}
+
+	// Set up out new info value.
+	info := Info{}
+
+	// Range over the groups and assign them based on index.
+	for i, groupString := range groups {
+		// Convert the index into a pnGroup enum.
+		group := pnGroup(i)
```

Comment:
if we're going to do it like this, I feel we shouldn't use iota in the declaration (because the enum values will change if someone inadvertently later on adds a new value in the middle of the group). Instead, may be best to declare the consts with integer values directly.

Reply:
Will do.